### PR TITLE
Index Filter

### PR DESCRIPTION
**Pull request by danrbailey**
* Index filters - random per leaf, hashed, bbox, level set, etc
* GR Primitive extension to display only a specific group
* OpenVDB Points Group SOP for filtering points

### CODE REVIEW

File: openvdb_points/tools/PointGroup.h [hunk 1]
```diff
@@ -248,6 +259,59 @@ struct SetGroupFromIndexOp
...
+                }
+            }
+            else {
+                IndexIter iter = leaf->beginIndexAll();
+                FilterIndexIter<IndexIter, FilterT> filterIndexIter(iter, filter);
+
+                for (; filterIndexIter; ++filterIndexIter) {
+                    group.set(*filterIndexIter, true);
+                }
+            }
```
Comment (Idclip) [inline openvdb_points/tools/PointGroup.h]: Should probably use an IndexAllIter here as this op is also handling the case where RequiresCoord is false and performs the calculation over all points, ignoring active states.

Comment (Idclip) [inline openvdb_points/tools/PointGroup.h]: Alternatively you could use itertraits and allow the user to template the filter on the active state type like the stuff we've done recently

File: openvdb_points/tools/IndexFilter.h [hunk 2]
```diff
@@ -199,6 +202,47 @@ class RandomLeafFilter
...
+    static AttributeHashFilter create(const LeafT& leaf, const Data& data) {
+        return AttributeHashFilter(data, AttributeHandle<IntType>::create(leaf.constAttributeArray(data.index)));
+    }
+
+    template <typename IterT>
+    bool valid(const IterT& iter) const {
+        const IntType id = mIdHandle->get(*iter);
+        const unsigned int seed = mData.seed + (unsigned int) id;
+        math::Rand01<double, RandGenT> randGen(seed);
+        return randGen() < double(mData.factor);
```
Comment (Idclip) [inline openvdb_points/tools/IndexFilter.h]: Why not store as a double

File: openvdb_points/tools/IndexFilter.h [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        const SeedCountPair& value = it->second;
+        const unsigned int seed = (unsigned int) value.first;
+        const Index total = leaf.pointCount();
+        const Index count = std::min(value.second, total);
+        return RandomLeafFilter(seed, count, total);
+    }
+
+    inline void next() const {
+        const_cast<int&>(mSubsetOffset)++;
+        const_cast<int&>(mNextIndex) =  mSubsetOffset >= mCount ?
```
Comment (Idclip) [inline openvdb_points/tools/IndexFilter.h]: Be nicer if these were mutable

File: openvdb_points/tools/IndexFilter.h (hunk 2 above)
Comment (danrbailey) [inline openvdb_points/tools/IndexFilter.h]: Good point, I'll change that.

File: openvdb_points/tools/PointGroup.h (hunk 1 above)
Comment (danrbailey) [inline openvdb_points/tools/PointGroup.h]: Yeah I agree, they should definitely be consistent, I'll take a look.

File: openvdb_points/tools/IndexFilter.h [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    // reserve and fill vector with ascending indices
+    values.reserve(m);
+    for (int i = 0; i < int(m); i++)    values.push_back(i);
+
+    // shuffle indices using random generator
+    RandGen<RandGenT, IntType> randGen(seed, m);
+    std::random_shuffle(values.begin(), values.end(), randGen);
+
+    // sort the subset of the indices vector that will be used
+    std::sort(values.begin(), values.begin() + n);
```
Comment (Idclip) [inline openvdb_points/tools/IndexFilter.h]: should values be resized here to n after the sort? Or before and then sorted?

Comment (danrbailey) [inline openvdb_points/tools/IndexFilter.h]: Well, there's a tradeoff here. If the array is resized, that means allocating a new array (of the new size), copying over the values, de-allocating the old array. This all takes time, so this method simply sorts the first n elements of the existing array, then leaves it with elements beyond that being unsorted. The idea is that the user should only ever use the first n elements, as these are the only ones guaranteed to be correct and the assumption is this data is only temporary anyway so the only benefit is marginally improved memory usage.

I did actually have a second method where it inserts random values into a set repeatedly until it hits the required number of values and switch to using this value if n < 10%. Much more efficient for small n, however there was a noticeable difference in the method changing when it crossed this threshold and I couldn't find a good way of eradicating this, so I removed it altogether.

File: openvdb_points/tools/PointGroup.h (hunk 1 above)
Comment (danrbailey) [inline openvdb_points/tools/PointGroup.h]: Think I might just switch this method over to use IndexOnIter. @Idclip, can you point me at the use of the itertraits method you're talking about here?

Comment (Idclip) [inline openvdb_points/tools/PointGroup.h]: Hmm I've taken another look and seen that this becomes slightly more tricky because the FilterIndexIter is templated on the indexIter type and its impossible to get the coord with a generic IndexIter. I was hoping to do something like this:

```
template<typename LeafNodeT, typename IndexIterT>
struct IndexIteratorTraits {};

template<typename LeafNodeT>
struct IndexIteratorTraits<LeafNodeT, typename LeafNodeT::IndexIter>
{
    typedef typename LeafNodeT::IndexIter IterT;
    static const bool HasCoord = false;
    static IterT begin(LeafNodeT& leaf) { return leaf.beginIndexAll(); }
};

template<typename LeafNodeT>
struct IndexIteratorTraits<LeafNodeT, typename LeafNodeT::IndexOnIter>
{
    typedef typename LeafNodeT::IndexOnIter IterT;
    static const bool HasCoord = true;
    static IterT begin(LeafNodeT& leaf) { return leaf.beginIndexOn(); }
};

template<typename LeafNodeT>
struct IndexIteratorTraits<LeafNodeT, typename LeafNodeT::IndexOffIter>
{
    typedef typename LeafNodeT::IndexOffIter IterT;
    static const bool HasCoord = true;
    static IterT begin(LeafNodeT& leaf) { return leaf.beginIndexOff(); }
};

template <typename PointDataTree, typename FilterT, typename IndexIterT = tools::IndexIter>
struct SetGroupByFilterOp
{
    typedef typename PointDataTree::LeafNodeType   LeafNodeT;

    void operator()(const typename LeafManagerT::LeafRange& range) const
    {
        for (typename LeafManagerT::LeafRange::Iterator leaf=range.begin(); leaf; ++leaf) 
        {
            ...

            if (FilterTraits<FilterT>::RequiresCoord && 
                IndexIteratorTraits<IndexIterT>::HasCoord) 
            {
                IndexIterT iter = IndexIteratorTraits<LeafNodeT, IndexIterT>(*leaf)::begin();
                ...
            }
            else if (FilterTraits<FilterT>::RequiresCoord && 
                     !IndexIteratorTraits<IndexIterT>::HasCoord) 
            {
                ...
            }
       }
   }
};

```
Wanted to add a static Coord method on the traits but this still suffers from the same limitation. I'm not sure the best solution here.

File: openvdb_points/tools/IndexFilter.h (hunk 4 above)
Comment (Idclip) [inline openvdb_points/tools/IndexFilter.h]: Resize only performs a reallocation if the size is greater than the capacity? It was my understanding that its the same as calling erase on the end elements as long as n < capacity()

File: openvdb_points/tools/IndexFilter.h [hunk 5]
```diff
@@ -243,6 +244,70 @@ class AttributeHashFilter
...
+        Data(const LevelSetGridT& _grid, const math::Transform& _transform,
+             const ValueT _min, const ValueT _max)
+            : accessor(_grid.getConstAccessor())
+            , levelSetTransform(_grid.transform())
+            , transform(_transform)
+            , min(_min)
+            , max(_max) { }
+
+        // not a reference to ensure const-accessor is unique per-thread
+        const typename LevelSetGridT::ConstAccessor accessor;
```
Comment (Idclip) [inline openvdb_points/tools/IndexFilter.h]: In the unit tests below, data is only constructed once, then passed as a reference or pointer everywhere. Are we expecting the user to construct data locally for each thread?

Comment (Idclip) [inline openvdb_points/tools/IndexFilter.h]: as a side note, I think some multi threaded unit tests would be really useful

File: openvdb_points/tools/PointGroup.h (hunk 1 above)
Comment (danrbailey) [inline openvdb_points/tools/PointGroup.h]: Tried a different approach to address this, would be great to get your feedback.

File: openvdb_points/tools/IndexFilter.h (hunk 4 above)
Comment (danrbailey) [inline openvdb_points/tools/IndexFilter.h]: Thanks, that's a very good point, just changed it!

File: openvdb_points/tools/IndexFilter.h [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+namespace tools {
+
+
+////////////////////////////////////////
+
+
+namespace index_filter_internal {
+
+template <typename RandGenT, typename IntType>
+struct RandGen
```
Comment (jbrd) [inline openvdb_points/tools/IndexFilter.h]: Am I right in saying that this struct only exists to allow you to pass a `RandInt` into `std::random_shuffle`? We're not currently doing anything with the argument in `operator ()` and yet `std::random_shuffle` expects the return value to be non-negative and less than the value of this argument. It may be worth adding an assert to document this constraint...

File: openvdb_points/tools/IndexFilter.h [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+{
+    RandGen(const unsigned int seed, const IntType length) : randGen(seed, 0, length-1) { }
+    IntType operator()(IntType) { return randGen(); }
+    openvdb::math::RandInt<IntType, RandGenT> randGen;
+};
+
+// generate a random subset of n indices from the range [0:m]
```
Comment (jbrd) [inline openvdb_points/tools/IndexFilter.h]: Should this be [0:m)?

File: openvdb_points/unittest/TestIndexFilter.cc [hunk 8]
```diff
@@ -0,0 +1,812 @@
...
+    }
+
+    { // all exclude
+        std::vector<Name> include;
+        std::vector<Name> exclude; exclude.push_back("all");
+        MultiGroupFilter::Data data(include, exclude);
+        MultiGroupFilter filter = MultiGroupFilter::create(*leaf, data);
+        IndexGroupIter iter(indexIter, filter);
+        CPPUNIT_ASSERT_EQUAL(iterCount(iter), size_t(0));
+    }
```
Comment (jbrd) [inline openvdb_points/unittest/TestIndexFilter.cc]: Would be good to add a test here to ensure the MultiGroupFilter works when both include and exclude are non-empty. For instance, including even but excluding first...

File: openvdb_points/tools/IndexFilter.h [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+                                                itEnd = data.exclude.end(); it != itEnd; ++it) {
+            if (!leaf.attributeSet().descriptor().hasGroup(*it))    continue;
+            exclude.push_back(new GroupHandle(leaf.groupHandle(*it)));
+        }
+        return MultiGroupFilter(include, exclude);
+    }
+
+    template <typename IterT>
+    bool valid(const IterT& iter) const {
+        for (HandleVector::const_iterator   it = mIncludeHandles.begin(),
```
Comment (jbrd) [inline openvdb_points/tools/IndexFilter.h]: Would be good to document the order of precedence of includes vs excludes

File: openvdb_points/tools/IndexFilter.h [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+        return true;
+    }
+
+private:
+    HandleVector mIncludeHandles;
+    HandleVector mExcludeHandles;
+}; // class MultiGroupFilter
+
+
+struct LeafMapData
```
Comment (jbrd) [inline openvdb_points/tools/IndexFilter.h]: Only seems to be used by RandomLeafFilter - can we move this into the RandomLeafFilter scope?

File: openvdb_points/tools/IndexFilter.h [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+
+private:
+    const Data* mData;
+    const AttributeHandle<openvdb::Vec3f>::Ptr mPositionHandle;
+}; // class BBoxFilter
+
+
+// Index filtering based on evaluating both sub-filters
+template <typename T1, typename T2, bool And = true>
+class BinaryFilter
```
Comment (jbrd) [inline openvdb_points/tools/IndexFilter.h]: Have you considered splitting this into an AndFilter and OrFilter to make its functionality a bit more explicit, especially in the default case where people will omit the third template argument? (at the cost of more code admittedly!)

File: openvdb_points/tools/PointGroup.h [hunk 12]
```diff
@@ -678,6 +744,75 @@ inline void setGroup(   PointDataTree& tree,
 ////////////////////////////////////////
 
 
+template <typename PointDataTree, typename FilterT>
+inline void setGroupByFilter(   PointDataTree& tree,
+                                const Name& group,
+                                const typename FilterT::Data filterData)
```
Comment (jbrd) [inline openvdb_points/tools/PointGroup.h]: Should filterData be a reference?

**Comment by jbrd** [comment]
That concludes my feedback...

File: openvdb_points/tools/IndexFilter.h (hunk 6 above)
Comment (danrbailey) [inline openvdb_points/tools/IndexFilter.h]: Good idea, added the assert.

File: openvdb_points/tools/IndexFilter.h (hunk 7 above)
Comment (danrbailey) [inline openvdb_points/tools/IndexFilter.h]: Ha, technically it would be! I'm using the non-mathematical, purely Python-style slicing syntax here. :)

File: openvdb_points/tools/IndexFilter.h (hunk 10 above)
Comment (danrbailey) [inline openvdb_points/tools/IndexFilter.h]: Yep, done.

File: openvdb_points/tools/IndexFilter.h (hunk 11 above)
Comment (danrbailey) [inline openvdb_points/tools/IndexFilter.h]: Think it's better to leave it as one class, just because it's less code and I don't have to define more cases for the trait.

File: openvdb_points/tools/PointGroup.h (hunk 12 above)
Comment (danrbailey) [inline openvdb_points/tools/PointGroup.h]: Yep, done.
